Extended validators to allow array values, dependent field validations and added new rules

xanish opened the pull request:
Hi,

I am planning to use the library on one of my projects. Thought I'd share some changes here as well to help out the community.

- Added support for array of values in rule parameters
  - The current ParameterizedRuleInterface is restrictive in terms of having to specify exact number of values which a rule would receive
  - The new ArrayParameterizedRuleInterface aims to relax the constraints and allow rules to make use of multiple parameter values
  - Added new rules like `in` and `required_if` using the above interface
- Added support for dependent field validations
  - The `passes()` function now receives the current validation row which allows for dependent field validation checks
  - Added new rule `required_if` which is based on dependent field validation
- Added new rules
  - alpha_num
  - in:foo,bar,baz
  - integer
  - numeric
  - required
  - required_if:field,value1,value2
- Following rules now silently pass if the field value is empty. Since we now have a explicit `required` rule, we can easily ignore empty cells unless required by user explicitly
  - `url`, `in`, `integer`, `numeric` and `alpha_num` will now return true for null / empty string values
- Added test cases for new rules, updated existing ones
- TODO: documentation for new rules

Let me know if any change is required.

xanish commented:
Hi, @hoshomoh, can you help me with this. Checks still haven't completed it seems.

hoshomoh commented:
This is quite the change. I will take some time this coming weekend to review. Great initiative.

xanish commented:
Made a few changes recently

- Added an optional trimming function to trim row values if needed
- Reverted Min, Max and Between to only work for Numbers as this would've caused issues since more information was needed to figure out whether we should check for length or the exact value of the field. (Ideally to make these generalised we would have to check if the attribute already has a number or integer validation then validate for value and if not then validate for length)
- Since I removed the Length checks from above rules added MinLength and MaxLength validators

hoshomoh commented on src/Rules/In.php:
```diff
@@ -0,0 +1,65 @@
+     * The allowed parameters should be tokenized string
+     * e.g :min, :max, :first, :last etc.
+     */
+    public function allowedParameters(): array
```
I am struggling to see why this is needed. The `$parameters` is already an array.

hoshomoh commented on src/Contracts/ArrayParameterizedRuleInterface.php:
```diff
@@ -0,0 +1,22 @@
+     * parsed in the same order and format as allowedParameters().
+     * This will aid in mapping our parameters to their placeholders.
+     */
+    public function parseParameterValues(array $parameters): array;
```
I think this is only useful when there a single parameter is allowed. Instead of doing this, I would rather update https://github.com/hoshomoh/CSVUtils/blob/34afb02e0e2506346ce429c2d7ff64bb7eadd5d9/src/Helpers/FormatsMessages.php#L101

```
protected function replaceParameterPlaceholder(
    string $message,
    array $allowedParameters,
    array $parameters
): string {
    $hasMultipleAllowedParameter = count($allowedParameters) > 1;
    $search = $hasMultipleAllowedParameter ? $allowedParameters : $allowedParameters[0];
    $replace = $hasMultipleAllowedParameter ? $parameters : implode(',', $parameters);

    return str_replace($search, $replace, $message);
}
```

xanish replied:
Yes, this works as well. I just wanted to keep it open in case someone needed multiple parameters.

hoshomoh commented on src/Helpers/FormatsMessages.php:
```diff
@@ -88,6 +89,14 @@ protected function makeReplacements(
             );
         }
 
+        if ($rule instanceof ArrayParameterizedRuleInterface) {
```
See my comment above

hoshomoh commented on src/Rules/Between.php:
```diff
@@ -19,20 +19,18 @@ public function allowedParameters(): array
         list($min, $max) = $parameters;
 
-        return $size >= (int) $min && $size <= (int) $max;
+        return +$value >= +$min && +$value <= +$max;
```
Prefer explicit casting with `(int)`

xanish replied:
I went with this because I wanted to keep the rules generalised for both floats and ints. If we want an explicit cast then we would need separate rules for int and float. Maybe I can add an additional method like cast($parameters) in the Between rule and extend it for BetweenFloat. Let me know your thoughts on this.

```php
protected function cast(array $parameters): array
{
    return array_map('intval', $parameters);
}
```

hoshomoh replied:
You are right I wasn't thinking of floats. We can keep it as is

hoshomoh commented on src/Rules/In.php:
```diff
@@ -0,0 +1,65 @@
+     * parsed in the same order and format as allowedParameters().
+     * This will aid in mapping our parameters to their placeholders.
+     */
+    public function parseParameterValues(array $parameters): array
```
See my comment above

hoshomoh commented on src/Rules/Max.php:
```diff
@@ -0,0 +1,36 @@
+    {
+        list($max) = $parameters;
+
+        return +$value <= +$max;
```
Prefer explicit casting `(int)`

hoshomoh commented on src/Rules/Min.php:
```diff
@@ -0,0 +1,36 @@
+    {
+        list($min) = $parameters;
+
+        return +$value >= +$min;
```
Prefer explicit casting `(int)`

hoshomoh commented on src/Rules/RequiredIf.php:
```diff
@@ -0,0 +1,63 @@
+     * parsed in the same order and format as allowedParameters().
+     * This will aid in mapping our parameters to their placeholders.
+     */
+    public function parseParameterValues(array $parameters): array
```
I am not sure why this was needed at all. Maybe I am not seeing something. My guess is this will be used like `["requiredIf:field,value"]`

xanish replied:
Yes, for my use case I had a couple of fields which had to be made required in case some target column had one of the values specified in :other_values.

hoshomoh replied:
I see. It seems we can't run away from supporting array parameters. Let's do the then. I will make a fix to support array parameters like so,  all rules can accept parameters in one of the following ways 

- `["requiredIf:field,value"]`
- `["requiredIf" => ["field" => "value"]`
- `["requiredIf" => ["field1" => "value1", "field2" => "value2"]`
- `["requiredIf" => ["field1" => ["value1", "value2"]]`

What do you think? I will try to push this over the weekend.

xanish replied:
Sounds good.

hoshomoh commented on src/Validator/Validator.php:
```diff
@@ -357,6 +370,13 @@ protected function passesParameterCheck($rule, array $parameters): bool
             return $parameterCount === $ruleParameterCount;
         }
 
+        if ($rule instanceof ArrayParameterizedRuleInterface) {
```
This no longer needed

hoshomoh commented on src/Validator/Validator.php:
```diff
@@ -97,6 +98,13 @@ class Validator
+     *
+     * @var bool
+     */
+    protected $trimCells = false;
```
```suggestion
    protected $shouldTrim = false;
```

hoshomoh commented on src/Validator/Validator.php:
```diff
@@ -407,4 +427,15 @@ protected function getValue(string $attribute)
+     *
+     * @return void
+     */
+    public function enableCellValueTrim(bool $trimValues = false)
```
```suggestion
    public function setShouldTrim(bool $shouldTrim = false)
```
